In [2]:
import pandas as pd #data processing, CSV file I/O (e.g. pd.read_csv)
import numpy as np # numerical calculation, linear algebra
import matplotlib.pyplot as plt # visulizing plot
import seaborn as sns # for visilizing plot

In [3]:
# Lets check the columns name in dataframe as all columns are vot visible
pd.set_option('display.max_columns', None)

In [1]:
df = pd.read_csv("DataCoSupplyChainDataset.csv", encoding="latin1")
df.head() # Showing first 5 rows in dataframe, we can set df.head(number_of_rows) where as number_of_rows for how many rows we want to see

NameError: name 'pd' is not defined

In [67]:
print(df.shape)

(180519, 53)


From the output, we can see that the table contains ***180519*** rows and ***53*** columns.

#### Now let’s try printing out the column names using: ***columns***

In [68]:
print(df.columns)

Index(['Type', 'Days for shipping (real)', 'Days for shipment (scheduled)',
       'Benefit per order', 'Sales per customer', 'Delivery Status',
       'Late_delivery_risk', 'Category Id', 'Category Name', 'Customer City',
       'Customer Country', 'Customer Email', 'Customer Fname', 'Customer Id',
       'Customer Lname', 'Customer Password', 'Customer Segment',
       'Customer State', 'Customer Street', 'Customer Zipcode',
       'Department Id', 'Department Name', 'Latitude', 'Longitude', 'Market',
       'Order City', 'Order Country', 'Order Customer Id',
       'order date (DateOrders)', 'Order Id', 'Order Item Cardprod Id',
       'Order Item Discount', 'Order Item Discount Rate', 'Order Item Id',
       'Order Item Product Price', 'Order Item Profit Ratio',
       'Order Item Quantity', 'Sales', 'Order Item Total',
       'Order Profit Per Order', 'Order Region', 'Order State', 'Order Status',
       'Order Zipcode', 'Product Card Id', 'Product Category Id',
       'Product De

#### We can use the ***info()*** method to output some general information about the dataframe:

In [69]:
print(df.info()) # Check the data information like which type of data the column contains

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180519 entries, 0 to 180518
Data columns (total 53 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   Type                           180519 non-null  object 
 1   Days for shipping (real)       180519 non-null  int64  
 2   Days for shipment (scheduled)  180519 non-null  int64  
 3   Benefit per order              180519 non-null  float64
 4   Sales per customer             180519 non-null  float64
 5   Delivery Status                180519 non-null  object 
 6   Late_delivery_risk             180519 non-null  int64  
 7   Category Id                    180519 non-null  int64  
 8   Category Name                  180519 non-null  object 
 9   Customer City                  180519 non-null  object 
 10  Customer Country               180519 non-null  object 
 11  Customer Email                 180519 non-null  object 
 12  Customer Fname                

# *Data Cleaning*

From the output of ***info*** we can see that ***shipping date (DateOrders)*** and ***order date (DateOrders)*** is showing ***object*** data type. It should be ***datetime*** data type. So, Let's make it

In [70]:
df['shipping date (DateOrders)'] = pd.to_datetime(df['shipping date (DateOrders)'], format='%m/%d/%Y %H:%M')
df['order date (DateOrders)'] = pd.to_datetime(df['order date (DateOrders)'], format='%m/%d/%Y %H:%M')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180519 entries, 0 to 180518
Data columns (total 53 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   Type                           180519 non-null  object        
 1   Days for shipping (real)       180519 non-null  int64         
 2   Days for shipment (scheduled)  180519 non-null  int64         
 3   Benefit per order              180519 non-null  float64       
 4   Sales per customer             180519 non-null  float64       
 5   Delivery Status                180519 non-null  object        
 6   Late_delivery_risk             180519 non-null  int64         
 7   Category Id                    180519 non-null  int64         
 8   Category Name                  180519 non-null  object        
 9   Customer City                  180519 non-null  object        
 10  Customer Country               180519 non-null  object        
 11  

Now ***shipping date (DateOrders)*** and ***order date (DateOrders)*** columns are showing correct data type.

In [71]:
# UDF for getting duplicate columns name
def getDuplicateColumns(df):
    duplicateColumnNames = set() # Define empty set
     
    for x in range(df.shape[1]): # Iterate through all the columns of DataFrame
        col = df.iloc[:, x]
        for y in range(x + 1, df.shape[1]): #Iterate through all the columns in DataFrame from (x + 1)th index to last index
            otherCol = df.iloc[:, y]
             
            # If two columns are equal then adding to the set
            if col.equals(otherCol):
                duplicateColumnNames.add(df.columns.values[y])
                 
    # Return list of unique column names whose contents are duplicates
    return list(duplicateColumnNames)

columns = getDuplicateColumns(df)
print(columns)
print("\nTotal duplicate columns: ", len(columns))

['Order Profit Per Order', 'Product Card Id', 'Customer Password', 'Product Category Id', 'Order Customer Id', 'Order Item Total', 'Product Price']

Total duplicate columns:  7


These columns are duplicate, So we can drop these colums. Let's drop those

In [72]:
df = df.drop(columns = getDuplicateColumns(df), inplace=False)
print(df.shape)

(180519, 46)


Earlier there was 53 columns, After removing ***7*** duplicate columns, Now it have 46 columns

In [73]:
# UDF for getting duplicate columns name
def getOneDistinctValueCols(df):
    duplicateColumnNames = set() # Define empty set
    uniqueColumnNames = {}
    
    for col in list(df.columns): # Iterate through all the columns of DataFrame
        listUniqueValue = df[col].unique()
        if (len(listUniqueValue) == 1): # If column contains only signle unqiue value then adding to the dictionary
            uniqueColumnNames[col] = listUniqueValue[0];
    
    # Return dict of signle unique column names along with values.
    return dict(uniqueColumnNames)

columns = getOneDistinctValueCols(df)
print(columns)
print("\nTotal columns: ", len(columns))

{'Customer Email': 'XXXXXXXXX', 'Product Description': nan, 'Product Status': 0}

Total columns:  3


From the above result we can see that ***3*** columns contains only one unique values. So, we can drop those columns. Let's do it

In [84]:
df = df.drop(columns = getOneDistinctValueCols(df).keys(), inplace=False)
print(df.shape)

(180519, 44)


Earlier there was 53 columns, After removing ***3*** duplicate columns, Now it have 43 columns

In [75]:
# Merge customer first name & last name in single column 
df["Customer Name"] = df["Customer Fname"].astype(str) + " " + df["Customer Lname"].astype(str)

In [85]:
df_a = df.drop(columns = ["Customer Fname", "Customer Lname"], inplace=False) # Drop Customer Fname & Customer Lname columns
print(df_a.shape)

(180519, 42)


### Now let's check whether there have any null values or not

In [86]:
print(df_a.isnull().sum())

Type                                  0
Days for shipping (real)              0
Days for shipment (scheduled)         0
Benefit per order                     0
Sales per customer                    0
Delivery Status                       0
Late_delivery_risk                    0
Category Id                           0
Category Name                         0
Customer City                         0
Customer Country                      0
Customer Id                           0
Customer Segment                      0
Customer State                        0
Customer Street                       0
Customer Zipcode                      3
Department Id                         0
Department Name                       0
Latitude                              0
Longitude                             0
Market                                0
Order City                            0
Order Country                         0
order date (DateOrders)               0
Order Id                              0


From the result, We can see there is ***Order Zipcode*** column contains null value. And if we think about this column it is irrelevant in data analysis. So, we can drop this column as well. 

In [91]:
df_b = df_a.drop(columns = ["Order Zipcode"], inplace=False)
print(df_b.isnull().sum())

Type                             0
Days for shipping (real)         0
Days for shipment (scheduled)    0
Benefit per order                0
Sales per customer               0
Delivery Status                  0
Late_delivery_risk               0
Category Id                      0
Category Name                    0
Customer City                    0
Customer Country                 0
Customer Id                      0
Customer Segment                 0
Customer State                   0
Customer Street                  0
Customer Zipcode                 3
Department Id                    0
Department Name                  0
Latitude                         0
Longitude                        0
Market                           0
Order City                       0
Order Country                    0
order date (DateOrders)          0
Order Id                         0
Order Item Cardprod Id           0
Order Item Discount              0
Order Item Discount Rate         0
Order Item Id       

Now, we can see there is ***no null values*** in the above columns

# *Univarite Analysis*

* Univariate analysis:- provides summary statistics for each field in the raw data set (or) summary only on one variable. Ex:- CDF,PDF,Box plot, Violin plot.(don't worry, will see below what each of them is)
* Bivariate analysis:- is performed to find the relationship between each variable in the dataset and the target variable of interest (or) using 2 variables and finding the relationship between them.Ex:-Box plot, Violin plot.
* Multivariate analysis:- is performed to understand interactions between different fields in the dataset (or) finding interactions between variables more than 2. Ex:- Pair plot and 3D scatter plot.